From 4ba953577be8765a90228cc33d7fcc1dd232a23b Mon Sep 17 00:00:00 2001
From: Kelly McLaughlin <kelly.mclaughlin@joyent.com>
Date: Fri, 8 Sep 2017 15:06:12 -0600
Subject: [PATCH] MANTA-3338 muskie leaked a connection Reviewed by: David
 Pacheco <dap@joyent.com> Approved by: Jared Morrow <jm@joyent.com>

---
 lib/server.js | 15 ++++++++++++++-
 1 file changed, 14 insertions(+), 1 deletion(-)

diff --git a/lib/server.js b/lib/server.js
index a1e58a6..b04eb4b 100644
--- a/lib/server.js
+++ b/lib/server.js
@@ -426,7 +426,20 @@ function createServer(options, clearProxy) {
         _audit(req, res, route, err);
     });
 
-    server.on('after', _audit);
+    server.on('after', function (req, res, route, err) {
+        _audit(req, res, route, err);
+
+        if ((req.method === 'PUT' || req.method === 'POST') &&
+            res.statusCode >= 500) {
+            /*
+             * An error occurred on a PUT or POST request, but there may still
+             * be incoming data on the request stream. Call resume() in order to
+             * dump any remaining request data so the stream emits an 'end' and
+             * the socket resources are not leaked.
+             */
+            req.resume();
+        }
+    });
 
     return (server);
 }
-- 
2.21.0

